src/slice.cr

```diff
@@ -744,11 +744,7 @@ struct Slice(T)
   # b # => Slice[3, 2, 1]
   # a # => Slice[3, 1, 2]
   # ```
-  def sort(&block : T, T -> U) : Slice(T) forall U
-    {% unless U <= Int32? %}
-      {% raise "expected block to return Int32 or Nil, not #{U}" %}
-    {% end %}
-
+  def sort(&block : T, T -> Int32?) : Slice(T)
```

Comment:
I wonder if this change would break something. I can't come up with a concrete example of a `U` that's a subtype of `Int32?`.

Reply:
`Int32` and `Nil` are subtypes, but they both work with both variants.

The macro solution was introduced in #6611 with the following comment by @asterite 

> Note: I replaced an `Int32?` restriction with `U forall U` to allow for `sort`'s block to be of type `Int32` or `Int32?`, for performance reasons. Ideally we should be able to do `U < Int32?` or some other condition over a free variable, but that's currently not supported. Eventually we can improve that code.

I don't know why that's more performant (and if that's still true).

Reply:
@asterite ? Why would it be more performant?

Reply:
The sort code raises an exception if the block returns `nil`. If we make the return type to be `Int32 | Nil`, checking for `nil` will always happen, regardless of whether the block can actually return `nil`. If we don't restrict the block's type and it only returns `Int32`, the check nil goes away.

That said, I'm not finding any performance difference now, or my benchmark is flawed, so it's probably fine to change this 🤷

Reply:
OK, let's leave it as is until we have better evidence.

Reply:
That is, @straight-shoota restore these lines for now.